tx/textual/valuerenderer/timestamp_test.go

```diff
@@ -0,0 +1,164 @@
+			err := rend.Format(context.Background(), protoreflect.ValueOf(tc.proto.ProtoReflect()), wr)
+			require.NoError(t, err)
+			require.Equal(t, tc.text, wr.String())
+
+			rd := strings.NewReader(tc.text)
+			val, err := rend.Parse(context.Background(), rd)
+			require.NoError(t, err)
+			msg := val.Message().Interface()
+			timestamp, ok := msg.(*tspb.Timestamp)
+			require.Truef(t, ok, "want Timestamp, got %T", timestamp)
+			require.True(t, proto.Equal(timestamp, &tc.proto))
+		})
+	}
+}
+
+// jsonTimestamp is a variant of the protobuf Timestamp that uses a string
+// to hold the int64 "seconds" field, for the benefit of our numerically-
+// impoverished friennds on the Javascript side.
+type jsonTimestamp struct {
+	Seconds string
+	Nanos   int32
+}
+
+// ToProto converts the json timestamp to a protobuf timestamp.
+func (jt jsonTimestamp) ToProto() (*tspb.Timestamp, error) {
+	seconds, err := strconv.ParseInt(jt.Seconds, 10, 64)
+	if err != nil {
+		return nil, err
+	}
+	return &tspb.Timestamp{Seconds: seconds, Nanos: jt.Nanos}, nil
+}
+
+// timestampJsonTest is the type of test cases in the testdata file.
+// Uses a custom unmarshaler since the JSON representation is a 2-element
+// array.
+type timestampJsonTest struct {
+	timestamp jsonTimestamp
+	text      string
+}
+
+var _ json.Unmarshaler = &timestampJsonTest{}
+
+// UnmarshalJSON implements the json.Unmarshaler interface
+func (t *timestampJsonTest) UnmarshalJSON(b []byte) error {
+	a := []interface{}{&t.timestamp, &t.text}
+	return json.Unmarshal(b, &a)
+}
+
+func TestTimestampJsonTestcases(t *testing.T) {
+	raw, err := os.ReadFile("../internal/testdata/timestamp.json")
+	require.NoError(t, err)
+
+	var testcases []timestampJsonTest
+	err = json.Unmarshal(raw, &testcases)
+	require.NoError(t, err)
+
+	for _, tc := range testcases {
+		ts, err := tc.timestamp.ToProto()
+		require.NoError(t, err)
+		rend := valuerenderer.NewTimestampValueRenderer()
+
+		wr := new(strings.Builder)
+		err = rend.Format(context.Background(), protoreflect.ValueOf(ts.ProtoReflect()), wr)
+		require.NoError(t, err)
+		require.Equal(t, tc.text, wr.String())
+
+		rd := strings.NewReader(tc.text)
+		val, err := rend.Parse(context.Background(), rd)
+		require.NoError(t, err)
+		msg := val.Message().Interface()
+		timestamp, ok := msg.(*tspb.Timestamp)
+		require.Truef(t, ok, "want Timestamp, got %T", timestamp)
+		require.True(t, proto.Equal(timestamp, ts))
+	}
+}
+
+func TestTimestampBadFormat(t *testing.T) {
+	rend := valuerenderer.NewTimestampValueRenderer()
+	wr := new(strings.Builder)
+	err := rend.Format(context.Background(), protoreflect.ValueOf(dur.New(time.Hour).ProtoReflect()), wr)
+	require.Error(t, err)
+}
+
+func TestTimestampBadParse(t *testing.T) {
+	for _, tc := range []struct {
+		name string
+		text string
+	}{
+		{name: "empty", text: ""},
+		{name: "whitespace", text: "   "},
+		{name: "garbage", text: "garbage"},
+		{name: "silly_americans", text: "11/30/2007"},
```

Comment:
One question I had for @webmaster128  was whether the test case JSON files should contain error cases too. I would say yes.

Reply:
In the .json tests I created I made some cases where `null` values mean an error occurred without going into detail which error that is.

However, since I only implement message -> text and not the other way around, I don't have any parsing logic.

Am I missing something?

Reply:
It would be good to have the portable test cases more comprehensive, but I don't think we can make everything JSON-driven, e.g. the parser test for a bad I/O reader.

Nevertheless, I think it would be good to have the JSON capture the core functionality. Does JS have a library for parsing the protobuf text format? If so, we could have a universal test case format with tests being JSON objects with the fields:

- "proto": text representation of the proto. Would need to make a test proto message with fields for all the scalars.
- "text": expected text rendering
- "error": boolean indicating that an attempt to format the proto should result in an error, and there will be no "text" field for these cases.
- I don't think we need to test error cases for parsing - we only need to ensure success of round trips from valid proto messages.

Each test case file would have an expectation of what sort of proto message the "proto" field should generate. Alternatively, we could use "A" as the universal test proto and name one of the fields to use. Then we'd have a single test driver for all types.

BTW, I've done this for the [CEL Language](https://github.com/google/cel-spec)
with data-driven [test files](https://github.com/google/cel-spec/blob/master/tests/simple/testdata/proto3.textproto).

Unfortunately, the protobuf support for JavaScript doesn't seem to implement textproto. I've found [one project](https://github.com/brotchie/protobuf-textformat) that provides such support, but it does not appear to be maintained.

Reply:
So I added an "Error" field to the test cases for testing both bad Format and bad Parse cases.

Reply:
I also find having JSON objects with field names better for readability (especially for newcomers) than arrays 👍